### PR TITLE
[FSDP][1/N] Move wrapper `ModuleWrapPolicy` to new path

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104427
* #104418
* #104409
* #104408
* #104407
* __->__ #104346

This PR is the first in refactoring the auto wrapping, only affecting `ModuleWrapPolicy` for wrapper `FullyShardedDataParallel`. The end goal is to improve the auto wrapping infra to support:
- Checking valid frozen parameters (uniform frozenness per FSDP)
- Checking valid shared parameters (shared parameters assigned to their lowest-common-ancestor module or higher)
- Writing auto wrapping policies that may take multiple passes over the module tree
- Specifying different FSDP kwargs per FSDP instance (instead of enforcing the same for all FSDP instances constructed via an auto wrap policy)

The way I envision achieving this is that, we decouple the actual "wrapping" (which is `_post_order_apply()` in this PR) from constructing the wrapping targets and kwargs (which is `target_module_to_kwargs` in this PR). In that way, a policy reduces to just constructing that latter `target_module_to_kwargs` mapping.

I do not personally recommend the size-based policy, but if we wanted to implement that under this new organization, the tracking of wrapped/nonwrapped numel should be done in the pass over the module tree prior to the actual "wrapping". This modularization keeps the actual "wrapping" part simple. 

The change to how `old_dtype` is handled is mainly to avoid keeping a reference to `_override_module_mixed_precision()` function closure in each hook and to allow the function to take in all module clases at once to return which ones actually got overridden for the downstream error message. (We can directly store the global state as a mapping.)

To-do in follow-ups (not in order):
- Add frozen parameter check before `_post_order_apply()`
- Add shared parameter check before `_post_order_apply()`
- Expose wrapping policy that allows per module / per module class kwarg customization (where any unspecified kwarg adopts the root's kwarg)

### CODE REVIEW

File: torch/distributed/fsdp/_utils.py [hunk 1]
```diff
@@ -1,47 +1,69 @@
...
+                return _apply_to_tensors(partial(cast_fn, torch.float32, module), args)
 
             def forward_post_hook(module, args, output):
-                nonlocal old_dtype
-                if old_dtype is not None:
-                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+                # NOTE: If the forward did not have any floating-point tensors,
+                # then the dtype will not be set for this module, and we do not
+                # upcast the dtype.
+                if module in _MODULE_TO_INP_DTYPE:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_utils.py:56]: so _MODULE_TO_INP_DTYPE generalizes old_dtype to be on a per-module basis?

File: torch/distributed/fsdp/wrap.py [hunk 2]
```diff
@@ -32,6 +34,104 @@
...
+        module: nn.Module,
+        module_name: str,
+        parent_module: Optional[nn.Module],
+    ):
+        for child_module_name, child_module in module.named_children():
+            if child_module not in visited_modules:
+                visited_modules.add(child_module)
+                _post_order_apply_inner(child_module, child_module_name, module)
+        optional_module = fn(module)
+        if optional_module is not None:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/wrap.py:64]: should we also assert it is an instance of nn.Module?

**Review by rohan-varma — APPROVED** [review]
LGTM

File: torch/distributed/fsdp/_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_utils.py:56]: Yep!

File: torch/distributed/fsdp/wrap.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/wrap.py:64]: Good point!

**Review by wanchaol — COMMENTED** [review]
meta comment on this PR: is this really a "move wrapper policy to the new path"? The PR added a lots of new logic, i.e. adding new post_order_apply, maybe the PR name should be more descriptive?

**Comment by awgu** [comment]
> meta comment on this PR: is this really a "move wrapper policy to the new path"? The PR added a lots of new logic, i.e. adding new post_order_apply, maybe the PR name should be more descriptive?

Good point! I wonder how we can fit more info given the title character limit though :/

File: torch/distributed/fsdp/wrap.py [hunk 3]
```diff
@@ -32,6 +34,104 @@
...
+            assert module_name != "", (
```
Comment (fegin) [inline torch/distributed/fsdp/wrap.py:65]: nit: `assert module_name, (`

File: torch/distributed/fsdp/wrap.py [hunk 4]
```diff
@@ -32,6 +34,104 @@
...
+                "Non-root modules should have their module name set but got "
+                f"an empty module name for {module}"
+            )
+            setattr(parent_module, module_name, optional_module)
```
Comment (fegin) [inline torch/distributed/fsdp/wrap.py:69]: `parent_module` is defined as `Optional` should we also check if it is an instance of `nn.Module` as well?

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**